Add rosa-authenticator infrastructure support

TGPSKI opened the pull request:
# Add rosa-authenticator support to terrascript_client

This MR adds support for multiple terraform resource creation via Terrascript -> Terraform.

This MR depends on MR #38690 in Gitlab, the design document documenting the use case and technical decisions that go into this change.

In addition, this MR will depend on updates to qontract-schemas (https://github.com/app-sre/qontract-schemas/pull/151), forcing default value requirements for associated yaml files. 

App-interface representation of new resources:

```
terraformResources:

- provider: rosa-authenticator
  account: appsrefrp01ugw1
  identifier: rosa-cognito
  domain: cognito.fedramp.devshift.net
  sms_role_ext_id: foobar
  defaults: /terraform/rosa-authenticator-1.yml
```

bkez322 commented on reconcile/utils/terrascript_client.py:
```diff
@@ -4451,3 +4465,190 @@ def populate_tf_resource_route53_zone(self, resource, namespace_info):
+
+        # https://registry.terraform.io/providers/hashicorp/aws/3.60.0/docs/resources/cognito_identity_provider
+
+        # What is the difference between values and self.get_values?
```
Kind of confusingly, it looks like some `pop_tf_resource_BLANK` function defs have a function parameter called `common_values`, and others call it `values`... but I think they're the same thing.

bkez322 replied:
But let's assume we're sticking to `common_values` for this function. More on this below.

bkez322 replied:
As of the writing of this comment, it's not clear to me if we need/want parameter-groups, but if we do, we want to create a separate terraform resource out of it, and add that terraform resource to the list of resources being provisioned. Example: https://github.com/app-sre/qontract-reconcile/blob/0cd5131b7fdccd685e1e8ab7d06740d574365048/reconcile/utils/terrascript_client.py#L1799-L1801

TGPSKI replied:
These won't be parameter groups, just defaults for usage with the providers I'm adding here. My exposure to parameter groups has been specific to RDS and related data stores - are PGs a wider concept than I think?

bkez322 replied:
"My exposure to parameter groups has been specific to RDS"

That's probably correct, I just wasn't 100% sure at the time of that comment. So yeah there should be no need for PGs.

bkez322 commented on reconcile/utils/terrascript_client.py:
```diff
@@ -4451,3 +4465,190 @@ def populate_tf_resource_route53_zone(self, resource, namespace_info):
+        # How do they relate to default files vs supplied parameters?
+        # I'm trying to use RDS as an example and I am getting confused with application of pg_values (parameter group)
+
+        pool_args = values.get("user_pool_properties", None)
```
```suggestion
        pool_args = common_values.get("user_pool_properties", None)
```

bkez322 commented on reconcile/utils/terrascript_client.py:
```diff
@@ -4451,3 +4465,190 @@ def populate_tf_resource_route53_zone(self, resource, namespace_info):
+        # I'm trying to use RDS as an example and I am getting confused with application of pg_values (parameter group)
+
+        pool_args = values.get("user_pool_properties", None)
+        pool_client_args = values.get("user_pool_client_properties", None)
```
```suggestion
        pool_client_args = common_values.get("user_pool_client_properties", None)
```

bkez322 commented on reconcile/utils/terrascript_client.py:
```diff
@@ -4451,3 +4465,190 @@ def populate_tf_resource_route53_zone(self, resource, namespace_info):
+        organization_name = "rosa-fedramp"
+        cognito_user_pool_arn = "foobar"
+
+        rest_api_args = values.get("rest_api_properties", None)
+        gateway_args = values.get("gateway_resource_properties", None)
+        gateway_method_args = values.get("gateway_method_properties", None)
+        gateway_authorizer_args = values.get("gateway_authorizer_properties", None)
+        integration_args = values.get("integration_properties", None)
```
```suggestion
        rest_api_args = common_values.get("rest_api_properties", None)
        gateway_args = common_values.get("gateway_resource_properties", None)
        gateway_method_args = common_values.get("gateway_method_properties", None)
        gateway_authorizer_args = common_values.get("gateway_authorizer_properties", None)
        integration_args = common_values.get("integration_properties", None)
```

bkez322 commented on reconcile/utils/terrascript_client.py:
```diff
@@ -4451,3 +4465,190 @@ def populate_tf_resource_route53_zone(self, resource, namespace_info):
+        # https://registry.terraform.io/providers/hashicorp/aws/3.60.0/docs/resources/cognito_identity_provider
+
+        # What is the difference between values and self.get_values?
+        # How do they relate to default files vs supplied parameters?
```
In app-interface namespace files, `defaults` and `parameter_group` both are expected to use a app-interface YML filepath as a value. Therefore I would expect to see the `self.get_values()` function to be used at any point in this code file when we need to read the data from those YML files, and I would expect to see `values.get()` or `common_values.get()` to be used if we're just checking for the presence of the key in the namespace file.

Example: https://gitlab.cee.redhat.com/service/app-interface/-/blob/master/data/services/insights/advisor/namespaces/stage-advisor-stage.yml#L71-76

bkez322 commented on reconcile/utils/terrascript_client.py:
```diff
@@ -4451,3 +4465,190 @@ def populate_tf_resource_route53_zone(self, resource, namespace_info):
+            # How does terraform DAG work in terrascript?
+            #
+            # For instance, I need to reference the ID of the user pool defined above in the user_pool_domain resource
+            # below. Does terrascript automatically respect the terraform DAG and supply this value
```
"Does terrascript automatically respect the terraform DAG and supply this value"

I believe the answer is basically yes. Terrascript doesn't do anything special, nor does it need to do. It just produces raw JSON format terraform files (which are equivalent to the more common HCL format terraform files), and terraform reads these JSON format files directly. At that point, it is terraform that is automatically handling the deciding on the most optimal order of resource provisioning, exactly the same as if you were providing Terraform with a normal HCL format terraform file.

bkez322 commented:
(Mentioned in the PR description, but linking to make it easy to find)

We'll want to update [this file](https://github.com/app-sre/qontract-schemas/blob/main/schemas/openshift/terraform-resource-1.yml) to enforce the the YML format of cognito and api-gateways.

TGPSKI commented on reconcile/utils/terrascript_client.py:
```diff
@@ -4451,3 +4465,190 @@ def populate_tf_resource_route53_zone(self, resource, namespace_info):
+        # How do they relate to default files vs supplied parameters?
+        # I'm trying to use RDS as an example and I am getting confused with application of pg_values (parameter group)
+
+        pool_args = values.get("user_pool_properties", None)
```
Good catch. Yes, this was my first attempt. Revising values / common_values / self.get_values in my next round, coming this evening.

dwelch0 commented on reconcile/utils/terrascript_client.py:
```diff
@@ -4460,3 +4475,204 @@ def populate_tf_resource_route53_zone(self, resource, namespace_info):
+
+        # FIXME
+        # I need to find an example of this pattern. Creating a dependent resource in an outside
+        # provider and referencing it here
```
Grasping at straws a little here but potentially could reference the user_pool similar to [this example](https://github.com/app-sre/qontract-reconcile/blob/7b062e0983b421916cf736d258ecc5e35c86fc15/reconcile/utils/terrascript_client.py#L761). String literal `aws_cognito_user_pool` and concatenate the identifier?

cubismod commented on reconcile/utils/terrascript_client.py:
```diff
@@ -4460,3 +4475,204 @@ def populate_tf_resource_route53_zone(self, resource, namespace_info):
+
+            api_gateway_lb_resource = aws_lb(
+                "lb",
+                name=f'{organization_name}-lb',
```
Where's `organization_name` coming from?

TGPSKI commented:
# Update 5/27/22

cc @maorfr @bkez322 @dwelch0 @cubismod @AlShv 

Perfomed a full update to bring terrascript implementation in line with the latest terraform prototype (https://gitlab.redhat-osd.local/service/terraform-ocm-cognito-api-gw)

Added these repos to house lambda functions:

- https://gitlab.redhat-osd.local/service/rosa-authenticator-lambda-pre-signup
- https://gitlab.redhat-osd.local/service/rosa-authenticator-lambda-pre-token

Added helper functions for pulling lambda funcitons into terraform-resources context:

- get_rosa_authenticator_zip
- download_rosa_authenticator_zip

Updated python representations for all of the following resources:

# iam.tf

- resource "aws_iam_role" "sms_role"
- resource "aws_iam_role" "lambda_role"

# lambda.tf

- resource "aws_lambda_function" "cognito_pre_signup"
- resource "aws_lambda_function" "cognito_pre_token_generation"
- resource "aws_lambda_permission" "cognito_pre_signup_permission"
- resource "aws_lambda_permission" "cognito_pre_token_generation_permission"

# cognito.tf

- resource "aws_cognito_user_pool" "pool"
- resource "aws_cognito_user_pool_domain" "userpool_domain"
- resource "aws_cognito_user_pool_client" "userpool_client"

# api-gateway.tf

- resource "aws_api_gateway_vpc_link" "vpc_link"
- resource "aws_api_gateway_rest_api" "gw_api"
- resource "aws_api_gateway_resource" "gw_resource_proxy"
- resource "aws_api_gateway_resource" "gw_resource_token"
- resource "aws_api_gateway_method" "gw_method_any"
- resource "aws_api_gateway_method" "gw_method_token_get"
- resource "aws_api_gateway_method_response" "gw_method_token_get_response"
- resource "aws_api_gateway_integration" "gw_integration_proxy"
- resource "aws_api_gateway_integration" "gw_integration_token"
- resource "aws_api_gateway_integration_response" "gw_integration_response_token"
- resource "aws_api_gateway_authorizer" "gw_authorizer"
- resource "aws_api_gateway_deployment" "gw_deployment"
- resource "aws_api_gateway_stage" "gw_stage"

Not implemented:

- resource "aws_cognito_user_pool_ui_customization" "userpool_ui"

Need to verify:

- s3 interactions with congnito

maorfr commented:
the only thing that seems missing is the lambda repos + automation? i see the path says `/repos/service/...`, which leads me to think this is only working locally for you?

TGPSKI commented:
> 

That's correct

TGPSKI commented:
@maorfr I made the lambda repos in FR gl already... ok?

TGPSKI commented:
# Update 6/10

Continuing to update and revise this MR as the upstream terraform repo is updated as well.

TGPSKI commented:
At my own risk, I used automatic formatting on save. Alas, this work is continued in a new MR: https://github.com/app-sre/qontract-schemas/pull/177
